src/main/java/org/broadinstitute/hellbender/utils/SATagBuilder.java

```diff
@@ -0,0 +1,294 @@
+package org.broadinstitute.hellbender.utils;
+
+
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SAMUtils;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.SAMRecordToGATKReadAdapter;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A builder class that expands functionality for SA tags. Each SATagBuilder is associated with a {@link GATKRead} at
+ * construction and supports various options of editing existing SA tags as well as adding new SA tags. NOTE: in order
+ * for updates to the SA tag for a given read to be added to the read, {@link SATagBuilder#writeTag()} must be called!
+ *
+ * When an SATagBuilder is constructed, it will automatically parse any existing SA tag strings on the read into
+ * processable units. In order to get the SA tag units as reads, simply call {@link SATagBuilder#getArtificialReadsBasedOnSATag(SAMFileHeader)}.
+ * Existing SA tag units can be cleared using {@link SATagBuilder#clear}.
+ *
+ * In order to add a supplementary read to a tag, simply call {@link SATagBuilder#addTag(GATKRead)} or {@link SATagBuilder#addTag(SATagBuilder)}
+ * and the handler will add a new section to its SA tag at the begining or the end of the SA string depending if the other
+ * read is supplementary or not. In order to edit a read in the Builder, a tool author must call {@link SATagBuilder#removeTag(GATKRead)}
+ * or {@link SATagBuilder#removeTag(SATagBuilder)} to remove the old tag and then {@link SATagBuilder#addTag(GATKRead)} to
+ * add the edited version of the read back in.
+ *
+ * Additionally, the static method {@link SATagBuilder#setReadsAsSupplemental(GATKRead, List)} can be used to set a group of reads as being
+ * supplementary to each other with correct SA tags. If there are existing SA tags on any of the reads they will be
+ * preserved in the operation.
+ */
+public class SATagBuilder {
+    private List<SARead> supplementaryReads;
```

Comment:
I think these can be made final with the possible exception of thisRead if you want to lazily cache that value.